samtranslator/model/sam_resources.py

```diff
@@ -2194,3 +2201,106 @@ def _construct_appsync_schema(self, api_id: Intrinsicable[str]) -> GraphQLSchema
+        sam_expect(
+            deltasync_properties.get("BaseTableTTL"), self.logical_id, "DynamoDBConfig.DeltaSyncConfig.BaseTableTTL"
+        ).to_be_a_string()
+
+        sam_expect(
+            deltasync_properties.get("DeltaSyncTableName"),
+            self.logical_id,
+            "DynamoDBConfig.DeltaSyncConfig.DeltaSyncTableName",
+        ).to_be_a_string()
+
+        sam_expect(
+            deltasync_properties.get("DeltaSyncTableTTL"),
+            self.logical_id,
+            "DynamoDBConfig.DeltaSyncConfig.DeltaSyncTableTTL",
+        ).to_be_a_string()
+
+        return cast(DeltaSyncConfigType, deltasync_properties)
+
+    def _parse_dynamodb_datasource(self, ddb_properties: Dict[str, Any]) -> DynamoDBConfigType:
+        ddb_config: DynamoDBConfigType = {}
+
+        ddb_config["TableName"] = sam_expect(
+            ddb_properties.get("TableName"),
+            self.logical_id,
+            "DynamoDBConfig.TableName",
+        ).to_be_a_string()
+
+        ddb_config["AwsRegion"] = ddb_properties["Region"] if "Region" in ddb_properties else ref("AWS::Region")
+
+        if "DeltaSync" in ddb_properties:
+            deltasync_properties = ddb_properties["DeltaSync"]
+            deltasync_config = self._parse_deltasync_properties(deltasync_properties)
+            ddb_config["DeltaSyncConfig"] = deltasync_config
+
+        if "UseCallerCredentials" in ddb_properties:
+            ddb_config["UseCallerCredentials"] = ddb_properties["UseCallerCredentials"]
+
+        if "Versioned" in ddb_properties:
+            ddb_config["Versioned"] = ddb_properties["Versioned"]
+
+        return ddb_config
+
+    def _validate_config_properties(self, datasource: DataSource) -> None:
+        # DataSourceConfig is quite large property so we require a lot of additional validation.
+        # The datasource object is modified by reference.
+        if self.Type == "AMAZON_DYNAMODB":
+            if not self.DynamoDBConfig:
+                raise InvalidResourceException(
+                    self.logical_id, "'DynamoDBConfig' must be defined when 'Type' is 'AMAZON_DYNAMODB'."
+                )
+            datasource.DynamoDBConfig = self._parse_dynamodb_datasource(self.DynamoDBConfig)
+
+    def _construct_appsync_datasource(self) -> DataSource:
+        datasource = DataSource(
+            logical_id=self.logical_id, depends_on=self.depends_on, attributes=self.resource_attributes
+        )
+
+        if self.Type not in SUPPORTED_DATASOURCES:
+            raise InvalidResourceException(self.logical_id, f"'{self.Type}' is not a supported data source type.")
+
+        datasource.Type = self.Type
+        datasource.Name = self.Name if self.Name else self.logical_id
+        datasource.Description = self.Description
+        datasource.ApiId = self.ApiId
+        # I will remove this ignore after, it will be fixed with pseudo-embedded connector implementation.
+        datasource.ServiceRoleArn = self.ServiceRoleArn  # type: ignore
```

Comment:
> type: ignore

Why?

Reply:
`ServiceRoleArn` is an optional property for the SAM resource, but is required on CFN resource.

This is because we are going to generate the role and policy if the customer does not define it. But that has not been implemented yet, so we're technically assigning an optional value to a required one which is a typing error. Once I create the connector change and push that, this value will always be defined so the `# type: ignore` will be removed.

I could change all the types to account for this now and then change them back with the connector PR but I don't think it's necessary since there's no customer impact at this time.